### PR TITLE
feat(parallelDockerUpdatecli) allow passing custom configuration to updatecli (such as a custom Docker Image)

### PR DESCRIPTION
**Pull request by dduportal**
This PR allow passing custom configuration to updatecli (such as a custom Docker Image).

Former syntax is still supported but with a warning.

Tested in https://github.com/jenkins-infra/docker-openvpn/pull/139

### CODE REVIEW

File: test/groovy/ParallelDockerUpdatecliStepTests.groovy [hunk 1]
```diff
@@ -163,10 +180,78 @@ class ParallelDockerUpdatecliStepTests extends BaseTest {
...
+    addEnvVar('BRANCH_IS_PRIMARY', 'true')
+    script.call(
+      imageName: testImageName,
+      updatecliConfig: [
+        containerMemory: anotherContainerMemory,
+      ],
+    )
+    printCallStack()
+
+    // Then we expect a successfull build
```
Comment (lemeurherve) [inline test/groovy/ParallelDockerUpdatecliStepTests.groovy:206]: No "When" above?

Comment (lemeurherve) [reply test/groovy/ParallelDockerUpdatecliStepTests.groovy:206]: Same for L238

File: vars/parallelDockerUpdatecli.groovy [hunk 2]
```diff
@@ -30,9 +41,14 @@ def call(userConfig = [:]) {
...
       withCredentials([string(credentialsId: finalConfig.credentialsId,variable: 'UPDATECLI_GITHUB_TOKEN')]) {
-        updatecli(action: 'diff', containerMemory: finalConfig.containerMemory)
+        // Merging the 2 maps - https://blog.mrhaki.com/2010/04/groovy-goodness-adding-maps-to-map_21.html
+        Map updatecliConfig = finalConfig.updatecliConfig << [action: 'diff']
+
+        updatecli(updatecliConfig)
         if (env.BRANCH_IS_PRIMARY) {
-          updatecli(action: 'apply', cronTriggerExpression: finalConfig.cronTriggerExpression, containerMemory: finalConfig.containerMemory)
+          // Merging the 2 maps - https://blog.mrhaki.com/2010/04/groovy-goodness-adding-maps-to-map_21.html
+          updatecliConfig = finalConfig.updatecliConfig << [action: 'apply', cronTriggerExpression: '@weekly']
```
Comment (lemeurherve) [inline vars/parallelDockerUpdatecli.groovy:50]:
```suggestion
          updatecliConfig = finalConfig.updatecliConfig << [action: 'apply', cronTriggerExpression: finalConfig.cronTriggerExpression]
```

Comment (lemeurherve) [reply vars/parallelDockerUpdatecli.groovy:50]: Or something like that (not hardcoded as '@weekly')

Comment (dduportal) [reply vars/parallelDockerUpdatecli.groovy:50]: mmmh, at first sight the suggestion is wrong, but your intent makes sense. let me try something to improve this. I'll update the thread here once pushed.

Comment (lemeurherve) [inline vars/parallelDockerUpdatecli.groovy:50]:
```suggestion
          updatecliConfig = finalConfig.updatecliConfig << [action: 'apply', cronTriggerExpression: finalConfig.updatecliConfig.cronTriggerExpression]
```

File: vars/parallelDockerUpdatecli.groovy [hunk 3]
```diff
@@ -30,9 +42,14 @@ def call(userConfig = [:]) {
...
       withCredentials([string(credentialsId: finalConfig.credentialsId,variable: 'UPDATECLI_GITHUB_TOKEN')]) {
-        updatecli(action: 'diff', containerMemory: finalConfig.containerMemory)
+        // Merging the 2 maps - https://blog.mrhaki.com/2010/04/groovy-goodness-adding-maps-to-map_21.html
+        Map updatecliConfig = finalConfig.updatecliConfig << [action: 'diff']
+
+        updatecli(updatecliConfig)
         if (env.BRANCH_IS_PRIMARY) {
-          updatecli(action: 'apply', cronTriggerExpression: finalConfig.cronTriggerExpression, containerMemory: finalConfig.containerMemory)
+          // Merging the 2 maps - https://blog.mrhaki.com/2010/04/groovy-goodness-adding-maps-to-map_21.html
+          updatecliConfig = finalConfig.updatecliConfig << [action: 'apply', cronTriggerExpression: defaultUpdatecliCronTriggerExpression]
```
Comment (lemeurherve) [inline vars/parallelDockerUpdatecli.groovy:51]: I fail to see how the cronTriggerExpression can be changed by the user with this.
I would have expected here something like `defaultConfig.updatecliConfig.cronTriggerExpression` or better yet, something like `finalConfig.updatecliConfig.cronTriggerExpression`

File: vars/parallelDockerUpdatecli.groovy (hunk 2 above)
Comment (dduportal) [reply vars/parallelDockerUpdatecli.groovy]: if `updatecliConfig` is null, then the suggestion triggers a java error stack (NullPointer)

File: vars/parallelDockerUpdatecli.groovy (hunk 3 above)
Comment (dduportal) [reply vars/parallelDockerUpdatecli.groovy:51]: look at the test :) 
Joke out:
```
parallelDockerUpdatecli([
  image 'foo',
  updatecliConfig: [
    cronTriggerExpression: '@daily',
  ],
])
```

File: test/groovy/ParallelDockerUpdatecliStepTests.groovy (hunk 1 above)
Comment (dduportal) [reply test/groovy/ParallelDockerUpdatecliStepTests.groovy]: should be better with last version, WDYT?

File: test/groovy/ParallelDockerUpdatecliStepTests.groovy [hunk 4]
```diff
@@ -163,9 +180,50 @@ class ParallelDockerUpdatecliStepTests extends BaseTest {
...
+    // When calling with the "parallelDockerUpdatecli" function with custom parameters, including legcay parameters for updatecli
+    addEnvVar('BRANCH_IS_PRIMARY', 'true')
+    script.call(
+      imageName: testImageName,
+      containerMemory: anotherContainerMemory,
+      cronTriggerExpression: anotherCronTriggerExpression,
+    )
+    printCallStack()
+
+    // Then we expect a successfull build
```
Comment (lemeurherve) [inline test/groovy/ParallelDockerUpdatecliStepTests.groovy:211]:
```suggestion
    // Then a successfull build is expected
```
To keep a passive (?) form

File: test/groovy/ParallelDockerUpdatecliStepTests.groovy [hunk 5]
```diff
@@ -163,9 +180,50 @@ class ParallelDockerUpdatecliStepTests extends BaseTest {
...
+    assertTrue(assertMethodCallContainsPattern('updatecli', 'action=apply'))
+    assertTrue(assertMethodCallContainsPattern('updatecli', 'cronTriggerExpression=' + anotherCronTriggerExpression))
+    assertTrue(assertMethodCallContainsPattern('updatecli', 'containerMemory=' + anotherContainerMemory))
+  }
+
+  @Test
+  void itRunsSuccessfullyWithOutdatedUpdatecliConfig() throws Exception {
+    def script = loadScript(scriptName)
+
+    // When calling with the "parallelDockerUpdatecli" function with custom parameters, including legcay parameters for updatecli
```
Comment (lemeurherve) [inline test/groovy/ParallelDockerUpdatecliStepTests.groovy:202]:
```suggestion
    // When the "parallelDockerUpdatecli" function is called with custom parameters, including legcay parameters for updatecli
```

Comment (lemeurherve) [reply test/groovy/ParallelDockerUpdatecliStepTests.groovy:202]: Same for L157

File: test/groovy/ParallelDockerUpdatecliStepTests.groovy [hunk 6]
```diff
@@ -132,21 +144,26 @@ class ParallelDockerUpdatecliStepTests extends BaseTest {
...
+    assertFalse(assertMethodCallContainsPattern('echo', 'WARNING:'))
   }
 
 
   @Test
   void itRunsSuccessfullyWithCustomParameters() throws Exception {
     def script = loadScript(scriptName)
 
-    // when calling with the "parallelDockerUpdatecli" function with custom parameters
+    // When calling with the "parallelDockerUpdatecli" function with custom parameters
```
Comment (lemeurherve) [inline test/groovy/ParallelDockerUpdatecliStepTests.groovy:157]:
```suggestion
    // When the "parallelDockerUpdatecli" function is called with custom parameters
```

**Review by lemeurherve — APPROVED** [review]

**Pull request merged**